### ISSUE TITLE
Inconsistency of internal sky light handling in subchunks

### ISSUE THREAD

**Issue by dktapps** [pmmp/PocketMine-MP#2533]
Currently both `SubChunk` and `EmptySubChunk` default to `15` for sky light on creation. This is wrong for several reasons.

When requesting a subchunk which is outside of the bounds of the chunk, an `EmptySubChunk` will be returned, which has all sky light as 15 (hardcoded). This causes bugs where "sky light" can propagate upwards through holes in the bedrock and light underground tunnels. This is not expected behaviour.

Additionally, this default `15` behaviour is undesirable for most non-empty subchunks, since these will rarely be free of blocks with opacity.

The rationale for empty subchunks having sky light all 15 was because of empty chunks above the surface which would usually be filled with 15 sky light. However, this "always default 15" is a bad solution which causes more problems than it solves.

**Comment by dktapps** [comment]
This problem has been fixed with respect to lighting updates, although grass growth at the top of the world still depends on this behaviour implicitly.

**Comment by dktapps** [comment]
For posterity, some reasons why it was done this way:

- Subchunks used to not always be allocated (they were stubbed by `EmptySubChunk`). When they were allocated, they were initialized with a default sky light level of 15, because it was assumed they would only ever be generated at the top of a chunk above the heightmap.
  - This assumption is obviously flawed, because large spaces in caves might result in empty, non-allocated subchunks, whose light levels would be incorrectly assumed to be 15.
  - This also created problems with sky light propagating upwards into caves through holes in the bedrock, since light was assumed to be 15 anywhere that `EmptySubChunk` was used, which included above and below the height limits.
  - This problem could have been addressed by assuming a default light of 0 instead, and then explicitly filling light of 15 if the subchunk was above the heightmap (although this could have been costly for performance).
- `EmptySubChunk` specifically was also used above and below the height limits, and since light is supposed to always be 15 at the top of the world, I cheaped out and made `EmptySubChunk` always have light level 15 too.

In newest versions of PM4, subchunks are always allocated whether they are empty or not, as are their light arrays (minus some trickery to save memory when they aren't used). This means that dynamic subchunk reallocation is no longer a problem, and the full extent of light-calculated chunks is always properly initialized.
